distance-method/distance_method.py

```diff
@@ -0,0 +1,981 @@
+        for line in phylip_f:
+            if debug:
+                sys.stdout.write("[DEBUG] %s" % line)
+            alignment_dist = line.strip().split()
+            if line.startswith(' '):
+                alignment_list.extend(alignment_dist)
+            else:
+                # new species alignment pairs
+                if alignment_list:
+                    for i in range(0, len(missing_species)):
+                        alignment_list.append(None)
+                    a = numpy.asarray(alignment_list[1:], dtype=float)
+                    a[ind_a] = numpy.nan
+                    ind_a += 1
+                    mean = numpy.nanmean(a)
+                    stdev = numpy.nanstd(a)
+                    # modify array with normalized values
+                    for x in numpy.nditer(a, op_flags=['readwrite']):
+                        x[...] = (x-mean)/stdev
+                    p[idx_p] = a
+                    idx_p += 1
+                    orig_order_labels.append(alignment_list[0])
+                alignment_list = alignment_dist
+
+    # add distance on final line
+    for i in range(0, len(missing_species)):
+        alignment_list.append(None)
+    a = numpy.asarray(alignment_list[1:], dtype=float)
+    a[ind_a] = numpy.nan
+    mean = numpy.nanmean(a)
+    stdev = numpy.nanstd(a)
+    for x in numpy.nditer(a, op_flags=['readwrite']):
+        x[...] = (x-mean)/stdev
+    p[idx_p] = a
+    orig_order_labels.append(alignment_list[0])
+
+    # add the missing species names to the labels array
+    bitvector_gene = 'I' * num_species
+    for species in missing_species:
+        orig_order_labels.append("%s_X" % species)
+        # indicate missing gene for current species
+        l = list(bitvector_gene)
+        l[int(species)] = 'O'
+        bitvector_gene = ''.join(l)
+
+    # update species set counts
+    if bitvector_gene not in species_set_dict:
+        species_set_dict[bitvector_gene] = 1
+    else:
+        species_set_dict[bitvector_gene] += 1
+
+    gene_bitvector_map[full_distance_matrix_offset] = bitvector_gene
+
+    # sort the distance matrix based on species names (S1, S2, S3 ..)
+    # in order to be consistent across all gene families
+    sorted_order_labels = sorted(orig_order_labels)
+    map_orig_sorted = {}
+    for idx_orig, label in enumerate(orig_order_labels):
+        idx_sorted = sorted_order_labels.index(label)
+        map_orig_sorted[idx_orig] = idx_sorted
+
+    # re-order rows and columns by ordered species name (0,1,2 ..)
+    p2 = numpy.zeros(shape=(num_species, num_species))
+    for idx_a, arr in enumerate(p):
+        t = numpy.zeros(shape=num_species)
+        for idx_b, el in enumerate(arr):
+            t[map_orig_sorted[idx_b]] = el
+        p2[map_orig_sorted[idx_a]] = t
+    del p
+
+    # add normalized distance matrix for current gene
+    # to full distance matrix
+    full_distance_matrix[full_distance_matrix_offset] = p2
+
+
+def cluster_distances(species_set_dict,
```

Comment:
I believe this function can be reworked so you do not have to iterate several times on `for bitvector in sorted_species_set:`